src/library_async.js

```diff
@@ -604,41 +635,92 @@ mergeInto(LibraryManager.library, {
+          Fibers.trampoline();
+        }
+
+        if (Asyncify.afterUnwind) {
+          Asyncify.afterUnwind();
+          Asyncify.afterUnwind = null;
+        }
+      }
+    },
+
+    instrumentExportSpecialized: function(original, arity, x) {
+      switch(arity) {
+        case 0: return function() {
+          Asyncify.exportCallStack.push(x);
+          try {
+            return original();
+          } finally {
+            if (ABORT) return;
+            Asyncify.finishInstrumentedExportCall(x);
+          }
+        }
+
+        case 1: return function(a1) {
+          Asyncify.exportCallStack.push(x);
+          try {
+            return original(a1);
+          } finally {
+            if (ABORT) return;
+            Asyncify.finishInstrumentedExportCall(x);
+          }
+        }
+
+        case 2: return function(a1, a2) {
+          Asyncify.exportCallStack.push(x);
+          try {
+            return original(a1, a2);
+          } finally {
+            if (ABORT) return;
+            Asyncify.finishInstrumentedExportCall(x);
+          }
+        }
+
+        default: return function() {
+          Asyncify.exportCallStack.push(x);
+          try {
+            return original.apply(null, arguments);
+          } finally {
+            if (ABORT) return;
+            Asyncify.finishInstrumentedExportCall(x);
+          }
+        }
+      }
+    },
+
     instrumentWasmExports: function(exports) {
       var ret = {};
       for (var x in exports) {
         (function(x) {
           var original = exports[x];
           if (typeof original === 'function') {
             ret[x] = function() {
-              Asyncify.exportCallStack.push(x);
-#if ASYNCIFY_DEBUG >= 2
-              err('ASYNCIFY: ' + '  '.repeat(Asyncify.exportCallStack.length) + ' try', x);
-#endif
-              try {
-                return original.apply(null, arguments);
-              } finally {
-                if (ABORT) return;
-                var y = Asyncify.exportCallStack.pop(x);
-                assert(y === x);
-#if ASYNCIFY_DEBUG >= 2
-                err('ASYNCIFY: ' + '  '.repeat(Asyncify.exportCallStack.length + 1) + ' finally', x);
-#endif
-                if (Asyncify.currData &&
-                    Asyncify.state === Asyncify.State.Unwinding &&
-                    Asyncify.exportCallStack.length === 0) {
-                  // We just finished unwinding.
-#if ASYNCIFY_DEBUG
-                  err('ASYNCIFY: stop unwind');
-#endif
-                  Asyncify.state = Asyncify.State.Normal;
-                  runAndAbortIfError(Module['_asyncify_stop_unwind']);
-                  if (Asyncify.afterUnwind) {
-                    Asyncify.afterUnwind();
-                    Asyncify.afterUnwind = null;
-                  }
-                }
-              }
+              var fun = Asyncify.instrumentExportSpecialized(original, arguments.length, x);
+              ret[x] = fun;
+              return fun.apply(null, arguments);
```

Comment:
Is this part of the change a general refactor?  Is so I wonder if we can land it separately?

I don't see `ret[x]` being assigned to in the old code.  was that a missed optimization?

Reply:
This is a part of an optimization, though it benefits general Asyncify usage as well as Fibers. The idea here is that the wrapper determines the function's arity when it's first called, and replaces itself with an arity-specific version that does not make use of `arguments`. This is inspired by #9922 which does the same thing on a larger scale, though not for Asyncify wrappers as far as I can tell.

I'm okay with reverting this change for now and perhaps letting #9922 handle it, or landing it later. It's kind of wonky anyway; I haven't investigated yet but I suspect it's the reason behind the failing tests. It does give a nice and measurable performance boost, though.

Reply:
I would say if this is separable and separately useful outside of fibres then it would be better to land this as a separate PR.